CalibTracker/SiStripChannelGain/src/APVGainHelpers.cc

```diff
@@ -0,0 +1,182 @@
+    if (id==6) return (int)TECDetId( det_id ).side();
+    return 0;
+}
+
+/** Brief Extract the subdetector side from a char * descriptor
+ * Return and integer whose coding is
+ *   0 - no side description can be applied
+ *   1 - for negative side
+ *   2 - for positive side
+ *
+ *   The char * descriptor is expected to have either "minus" or "plus"
+ *   string to specify the sign. If no sign spec is found 0 is returned.
+ */
+int APVGain::subdetectorSide(const char* tag) {
+    std::string t = std::string(tag);
+    std::size_t m = t.find("minus");
+    std::size_t p = t.find("plus");
+    if (m!=std::string::npos) return 1;
+    if (p!=std::string::npos) return 2;
+    return 0;
+}
+
+/** Brief Extract the detector plane position from a DetId.
+ * Return an integer that represent the detector plane where the module sits.
+ * For the barrel detectors (TIB and TOB) the detector plane is the layer, e.g.
+ * ranging from 1 to 4 in the TIB and from 1 to 6 in the TOB. For the endcap 
+ * detectors the detector plane is the wheel number with a sign in front to 
+ * tell in which side the wheel is sitting.
+ */
+int APVGain::subdetectorPlane(uint32_t det_id) {
+    if( APVGain::subdetectorId(det_id)==3 )      return TIBDetId( det_id ).layer();
+    else if( APVGain::subdetectorId(det_id)==4 ) return (2*(int)TIDDetId( det_id ).side()-3)*(int)TIDDetId( det_id ).wheel();
+    else if( APVGain::subdetectorId(det_id)==5 ) return TOBDetId( det_id ).layer();
+    else if( APVGain::subdetectorId(det_id)==6 ) return (2*(int)TECDetId( det_id ).side()-3)*(int)TECDetId( det_id ).wheel();
+    return 0;
+};
+
+/** Brief Extract from a char * the subdetector type.
+ * Return an integer whioch is the detector plane where the module sits.
+ * The char * string is expected to have the subdetector plane put at its
+ * end after an "_" char.
+ */
+int APVGain::subdetectorPlane(const char* tag) {
+   std::string s = std::string(tag);
+   std::size_t p = (s.find("layer")!=std::string::npos)? s.find("layer") : s.find("wheel");
+   if (p!=std::string::npos) {
+       std::size_t start = s.find("_",p+1) + 1;
+       std::size_t stop  = s.find('_',start);
+       std::string plane = s.substr(start,stop-start);
+       return atoi( plane.c_str());
+   }
+   return 0;
+};
+
+
+/** Brief Fetch the Monitor Element corresponding to a DetId.
+ *  */
+std::vector<MonitorElement*> APVGain::FetchMonitor(std::vector<MonitorElement*> histos, uint32_t det_id) {
+    std::vector<MonitorElement*> found = std::vector<MonitorElement*>();
+    std::vector<MonitorElement*>::iterator it= histos.begin();
+    while (it!=histos.end()) {
+        std::string tag = (*it)->getName();
+        int sId    = APVGain::subdetectorId((uint32_t)det_id);
+        int sPlane = APVGain::subdetectorPlane((uint32_t)det_id);
+        int sSide  = APVGain::subdetectorSide((uint32_t)det_id);
+
+        bool match = (APVGain::subdetectorId(tag.c_str())==0 || APVGain::subdetectorId(tag.c_str())==sId) &&
```

Comment:
these attributes were known at construction time - i would suggest saving them in a small cache (or just make the structure of the vector defined in monHnames more public than it is